runtime/compiler/optimizer/HotFieldMarking.cpp

```diff
@@ -0,0 +1,173 @@
+#include "il/Node.hpp"
+#include "il/Node_inlines.hpp"
+#include "il/SymbolReference.hpp"
+#include "il/TreeTop.hpp"
+#include "infra/ILWalk.hpp"
+#include "optimizer/Optimizer.hpp"
+#include "optimizer/Optimization.hpp"
+#include "optimizer/Optimization_inlines.hpp"
+#include "optimizer/OptimizationManager.hpp"
+
+#include <map>
+
+struct SymStats
+   {
+   int32_t _count;
+   int32_t _score;
+   int32_t _fieldNameLength;
+   char* _fieldName;
+   int32_t _fieldSigLength;
+   char* _fieldSig;
+   TR_OpaqueClassBlock *_clazz;
+   SymStats(int32_t count, int32_t score, int32_t fieldNameLength, char* fieldName, int32_t fieldSigLength, char* fieldSig, TR_OpaqueClassBlock *clazz) :
+      _count(count), _score(score), _fieldNameLength(fieldNameLength), _fieldName(fieldName), _fieldSigLength(fieldSigLength), _fieldSig(fieldSig), _clazz(clazz) {}
+   };
+
+typedef TR::typed_allocator<std::pair<TR::Symbol *, SymStats *>, TR::Region&> SymAggMapAllocator;
+typedef std::less<TR::Symbol *> SymAggMapComparator;
+typedef std::map<TR::Symbol *, SymStats *, SymAggMapComparator, SymAggMapAllocator> SymAggMap;
+
+static int32_t getReducedFrequencySum(int32_t currentValue, int32_t newFrequency)
+   {
+   return (currentValue + newFrequency);
+   }
+
+static int32_t getReducedFrequencyAverage(int32_t currentValue, int32_t count, int32_t newFrequency)
+   {
+   return ((currentValue * count) + newFrequency) / (count + 1);
+   }
+
+int32_t TR_HotFieldMarking::perform()
+   {
+   if (!jitConfig->javaVM->memoryManagerFunctions->j9gc_hot_reference_field_required(jitConfig->javaVM))
+      {
+      if (trace())
+         traceMsg(comp(), "Skipping hot field marking since dynamic breadth first scan ordering is disabled\n");
+      return 0;
+      }
+
+   SymAggMap stats(SymAggMapComparator(), comp()->trMemory()->currentStackRegion());
+   TR::Block *block = NULL;
+   for (TR::PostorderNodeIterator it(comp()->getStartTree(), comp()); it != NULL; ++it)
+      {
+      TR::Node * const node = it.currentNode();
+      if (node->getOpCodeValue() == TR::BBStart)
+         {
+         block = node->getBlock();
+         }
+      else if ((node->getOpCode().isLoadIndirect() || node->getOpCode().isStoreIndirect())
+               && node->getOpCode().hasSymbolReference()
+               && !node->getSymbolReference()->isUnresolved()
+               && node->getSymbolReference()->getSymbol()->isShadow()
+               && !node->isInternalPointer()
+               && !node->getOpCode().isArrayLength()
+               && node->getSymbolReference()->getSymbol()->isCollectedReference()
+              )
+         {
+         TR::SymbolReference *symRef = node->getSymbolReference();
+         if (symRef->getCPIndex() >= 0 && !symRef->getSymbol()->isArrayShadowSymbol())
+            {
+            auto itr = stats.find(symRef->getSymbol());
+            if (itr != stats.end())
+               {
+               if(TR::Options::getReductionAlgorithm(TR_HotFieldReductionAlgorithmSum))
```

Comment:
does this if need to be here? would it make more sense just to have the reductions all having the same signature and being static functions in the file so we could just assign a function pointer to point to one of the methods based on the configured algorithm? might make it easier to add other algorithms in the future or to even have the GC pass the reduction function in if it isn't embedded in the opt code directly...